CesiumRasterOverlays/src/UrlTemplateRasterOverlay.cpp

```diff
@@ -0,0 +1,189 @@
+#include <CesiumRasterOverlays/QuadtreeRasterOverlayTileProvider.h>
+#include <CesiumRasterOverlays/UrlTemplateRasterOverlay.h>
+#include <CesiumUtility/Uri.h>
+
+using namespace CesiumAsync;
+using namespace CesiumGeometry;
+using namespace CesiumGeospatial;
+using namespace CesiumUtility;
+
+namespace CesiumRasterOverlays {
+
+namespace {
+static bool caseInsensitiveCompare(std::string_view lhs, std::string_view rhs) {
+  return lhs.size() == rhs.size() &&
+         std::equal(lhs.begin(), lhs.end(), rhs.begin(), [](char a, char b) {
+           return std::tolower(a) == std::tolower(b);
+         });
+}
+} // namespace
+
+class UrlTemplateRasterOverlayTileProvider final
+    : public QuadtreeRasterOverlayTileProvider {
+public:
+  UrlTemplateRasterOverlayTileProvider(
+      const IntrusivePointer<const RasterOverlay>& pOwner,
+      const CesiumAsync::AsyncSystem& asyncSystem,
+      const std::shared_ptr<IAssetAccessor>& pAssetAccessor,
+      std::optional<Credit> credit,
+      const std::shared_ptr<IPrepareRasterOverlayRendererResources>&
+          pPrepareRendererResources,
+      const std::shared_ptr<spdlog::logger>& pLogger,
+      const CesiumGeospatial::Projection& projection,
+      const CesiumGeometry::QuadtreeTilingScheme& tilingScheme,
+      const CesiumGeometry::Rectangle& coverageRectangle,
+      const std::string& url,
+      const std::vector<IAssetAccessor::THeader>& headers,
+      uint32_t width,
+      uint32_t height,
+      uint32_t minimumLevel,
+      uint32_t maximumLevel)
+      : QuadtreeRasterOverlayTileProvider(
+            pOwner,
+            asyncSystem,
+            pAssetAccessor,
+            credit,
+            pPrepareRendererResources,
+            pLogger,
+            projection,
+            tilingScheme,
+            coverageRectangle,
+            minimumLevel,
+            maximumLevel,
+            width,
+            height),
+        _url(url),
+        _headers(headers) {}
+
+  virtual ~UrlTemplateRasterOverlayTileProvider() {}
+
+protected:
+  virtual CesiumAsync::Future<LoadedRasterOverlayImage> loadQuadtreeTileImage(
+      const CesiumGeometry::QuadtreeTileID& tileID) const override {
+
+    LoadTileImageFromUrlOptions options;
+    options.rectangle = this->getTilingScheme().tileToRectangle(tileID);
+    options.moreDetailAvailable = tileID.level < this->getMaximumLevel();
+
+    const GlobeRectangle unprojectedRect =
+        unprojectRectangleSimple(this->getProjection(), options.rectangle);
+
+    const std::string substitutedUrl = Uri::substituteTemplateParameters(
+        this->_url,
+        [this, tileID, &rectangle = options.rectangle, &unprojectedRect](
+            const std::string& placeholder) {
+          if (caseInsensitiveCompare(placeholder, "x")) {
```

Comment:
I think it would be worthwhile to build a std::map (with a case-insensitive key comparer) once and then use it for much faster lookups. This is what the WMTS tile provider does, minus the case-insensitivity.